hw3/heat.cpp

```diff
@@ -0,0 +1,217 @@
+#include <cmath>
+#include <fstream>
+#include <hip/hip_runtime.h>
+#include <iostream>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string>
+#include <vector>
+#include "mat.h"
+#include "CImg.h"
+
+#define M 1024
+#define BLOCK_SIZE 16
+using GMat3 = GpuMatrix3;
+using CMat3 = CpuMatrix3;
+using GMat = GpuMatrix;
+using CMat = CpuMatrix;
+using namespace cimg_library;
+
+__global__ void BlurrKernel(const unsigned char *in, int width, int height, const GMat Filter, unsigned char *out);
+__global__ void L2Kernel(const unsigned char *blurImg, int widht, int height, const unsigned char *origImg, GMat L2);
+__global__ void reduce(GMat dOut, GMat dIn);
+
+
+__global__ void BlurrKernel(const unsigned char *in, int width, int height, const GMat Filter, unsigned char *out)
+{
+   //Image width by height
+   int row = blockIdx.y * blockDim.y + threadIdx.y;
+   int col = blockIdx.x * blockDim.x + threadIdx.x;
+   float3 pixelValue;
+   pixelValue.x = 0.0f;
+   pixelValue.y = 0.0f;
+   pixelValue.z = 0.0f;
+   //Edge of image
+   if(row == 0 || row == height-1 || col == 0 || col == width-1)
+   {
+      //Load rgb pixel values for: input[row][col]
+      pixelValue.x = (float)(in[row * width + col]);
+      pixelValue.y = (float)(in[(row + height) * width + col]);
+      pixelValue.z = (float)(in[(row + height*2) * width + col]);
+       
+   }
+   else
+   {
+      for(int u = -1; u < 2; u ++)
+      {
+         for(int v = -1; v < 2; v++)
+         {
+            //Enumerate rgb pixel values for: input[row+u][col+v]*Filter[u+1][v+1]
+            pixelValue.x += (float)(in[(row+u) * width + (col+v)])*Filter.elements[(u+1)*Filter.width+(v+1)];
+            pixelValue.y += (float)(in[((row+u) + height) * width + (col+v)])*Filter.elements[(u+1)*Filter.width+(v+1)];
+            pixelValue.z += (float)(in[((row+u) + height*2) * width + (col+v)])*Filter.elements[(u+1)*Filter.width+(v+1)];
+         }
+       }
+
+   }
+    
+   //Load rgb values into output[row][col]
+   out[row * width + col] = pixelValue.x;
+   out[(row + height) * width + col] = pixelValue.y;
+   out[(row + height*2) * width + col] = pixelValue.z;
```

Comment:
You'll want to cast as uint here if you're casting to float before. The compiler casting may not be reliable. 

You'll also want to cap at 255. 
`out[row * width + col] = pixelValue.x > 255 ? 255 : pixelValue.x;`